x-pack/plugins/security_solution/public/management/services/exceptions_list/exceptions_list_api_client.ts

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import {
+  CreateExceptionListItemSchema,
+  CreateExceptionListSchema,
+  ExceptionListItemSchema,
+  ExceptionListSummarySchema,
+  FoundExceptionListItemSchema,
+  UpdateExceptionListItemSchema,
+} from '@kbn/securitysolution-io-ts-list-types';
+import { EXCEPTION_LIST_ITEM_URL, EXCEPTION_LIST_URL } from '@kbn/securitysolution-list-constants';
+import { HttpStart } from 'kibana/public';
+
+/**
+ * A generic class to be used for each artifact type.
+ * It follow the Singleton pattern.
+ * Please, use the getInstance method instead of creating a new instance when using this implementation.
+ */
+export class ExceptionsListApiClient {
```

Comment:
Suggestion: Make the class a Generic that can derive the List ID from the constructor, and then use that to narrow the input type and output types of the class methods to that list id. Example:


```typescript
interface GetResponse {
  id: string;
  list_id: string;
}

// Just a utility type to narrow the `list_id` to the given value
type NarrowToList<T extends object, ListId extends string> = T & { list_id: ListId };
// Or maybe: type NarrowToList<T extends object, ListId extends string> = Omit<T, 'list_id'> & { list_id: ListId };

export class ExceptionsListApiClient<L extends string> {
  constructor(private readonly listId: L) {}

  get(): NarrowToList<GetResponse, L> {
    return { id: 'sjkj', list_id: this.listId };
  }
  
  put(data: NarrowToList<GetResponse, L>) {} 
  
}

const c = new ExceptionsListApiClient('MY_LIST_ID');
const response = c.get();

response.list_id = 'one'; // ERROR: TS2322: Type '"one"' is not assignable to type '"MY_LIST_ID"'.

c.put({ id: 'one', list_id: 'some value'}); // Error: TS2322: Type '"some value"' is not assignable to type '"MY_LIST_ID"'.

```

It would also ensure that an instance of the class for a list is not attempted to be used for some other list type

Reply:
Does that means having just a generic class and remove the artifact specific implementations (`HostIsolationExceptionsApiClient`, `TrustedAppsApiClient`, `EventFiltersApiClient` )? If that the case, I thought having their own implementations would allow us to add hooks for custom validations or custom actions we want to do depending on the artifact type.

Reply:
Good point. Yes and no. Yes, one could just use this one generic class, but for the reasons you mentioned, I think we should crate sub-classes of this one for artifact specific API interface

Reply:
So we are on the same page then, I included the artifact specific API interface in this pr 🙂